File: algorithms/Ruby/Arrays/two_sum.rb
# frozen_string_literal: true

# Problem Statement: given an array of integers nums and an integer target,
# return indices of the two numbers such that they add up to target.
#
# You may assume that each input would have exactly one solution,
# and you may not use the same element twice.
#
# Input: An array of integers and a target (int)
# Output: array of indexes of len(2) with sum of element at that index equal to target or nil
#
# Time: O(n)
# Space: O(n)
module Arrays
  def self.two_sum(nums, target)
    complement_hash = {}

    nums.each_with_index do |num, index|
      complement = target - num

      return [complement_hash[complement], index] if complement_hash.key?(complement)

      complement_hash[num] = index
    end

    nil
  end
end

RSpec.describe Arrays do
  describe '.two_sum' do
    it 'returns the correct indices' do
      expect(Arrays.two_sum([2, 7, 11, 15], 9)).to eq([0, 1])
      expect(Arrays.two_sum([3, 2, 4], 6)).to eq([1, 2])
      expect(Arrays.two_sum([3, 3], 6)).to eq([0, 1])
    end

    it 'returns nil if no solution is found' do
      expect(Arrays.two_sum([1, 2, 3, 4], 9)).to be_nil
      expect(Arrays.two_sum([], 5)).to be_nil
    end
  end
end

File: algorithms/CPlusPlus/Arrays/two-sum.cpp
//Description :c++ solution to find all the unique two indices in an array whose sum is equal to the give number in O(n) complexity.
//using this method we will be able to find all the unique pairs
//Time Complexity : O(n), where n is the array size
//Space Complexity : O(n),map data structure will be used.
//sample
//input 1: nums = [2,7,11,15], target = 9
//output 1: [0,1]
//explanation: Because nums[0] + nums[1] == 9, we return [0, 1].

//input 2: nums = [3,2,4], target = 6
//output 2: [1,2]
//explanation: Because nums[1] + nums[2] == 6, we return [1, 2].

#include<iostream>
#include<vector>
#include<map>

using namespace std;

vector<int> twoSum(vector<int>& nums, int target) {
    map<int,int> mp;
    vector<int> vec;
    for(int i=0;i<nums.size();i++)
    {
        int temp = target - nums[i];
        if(mp.find(temp)==mp.end()){
            mp.insert({nums[i] ,i});
        }
        else{
            vec.push_back(mp[temp]);
            vec.push_back(i);
        }
    }
    return vec;
}

int main() {
    //you can take array and target as an input from user.
    vector<int> num = {2,7,11,15};
    vector<int> ans;
    int target = 9;
    ans = twoSum(num, target);
    for(int i=0;i<ans.size();i++){
        cout<<ans[i]<<" ";
    }
    cout<<endl;
    return 0;
}


